### PR TITLE
Allow specifying storageclass defined in the host cluster

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
related issues: https://github.com/harvester/harvester/issues/3052

Now we do not assign the default StorageClass for harvester-csi-driver.
Also, parse the parameter key `hostStorageClass` to a specific `StorageClass` on PVC creating.

Test Plan is described as below:

**Preparation:** 

**NOTE**: The preparation would be done on the host harvester cluster.
Please create the new clusterrole `harvesterhci.io:csi-driver` by following manifest:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    app.kubernetes.io/component: apiserver
    app.kubernetes.io/name: harvester
    app.kubernetes.io/part-of: harvester
  name: harvesterhci.io:csi-driver
rules:
- apiGroups:
  - storage.k8s.io
  resources:
  - storageclasses
  verbs:
  - get
  - list
  - watch
```

Then create the clusterrolebinding to associate with this. You could refer to the following manifest.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: <guest cluster namespace>-<guest cluster name>
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: harvesterhci.io:csi-driver
subjects:
- kind: ServiceAccount
  name: <your guest cluster name>
  namespace: <your guest cluster namespace>
```

**Test cases**:
- Verify the Default StorageClass behavior
1. Create the volume `vol001` with the default StorageClass on the Guest Cluster
2. Check the `vol001` is created with the Default StorageClass on the Host Harvester Cluster. 

- Verify the user-defined StorageClass behavior
1. Create a new StorageClass `test` on the Host Harvester Cluster.
2. Create a new StorageClass using the following manifest on the Guest Cluster.

    ``` yaml
    allowVolumeExpansion: true
    apiVersion: storage.k8s.io/v1
    kind: StorageClass
    metadata:
      name: harvester-SC-test
    provisioner: driver.harvesterhci.io
    reclaimPolicy: Delete
    volumeBindingMode: Immediate
    parameters:
      hostStorageClass: test
    ```

3. Create the volume `vol002` with the above StorageClass on the Guest Cluster.
4. Check the `vol002` is created with the `test` StorageClass on the Host Harvester Cluster.

- Verify the non-exists StorageClass behavior
1. Create a new StorageClass using the following manifest on the Guest Cluster.

    ``` yaml
    allowVolumeExpansion: true
    apiVersion: storage.k8s.io/v1
    kind: StorageClass
    metadata:
      name: harvester-SC-test-non-exist
    provisioner: driver.harvesterhci.io
    reclaimPolicy: Delete
    volumeBindingMode: Immediate
    parameters:
      hostStorageClass: non-storageclass
    ```

2. Create the volume `vol003` with the above StorageClass on the Guest Cluster.
3. Check the `vol002` is on the status `pending` and you can see some error log on the harvester-csi-driver.

### CODE REVIEW

**Review by bk201 — APPROVED** [review]
LGTM, works with RKE2.

**Review by guangbochen — COMMENTED** [review]
There is no valid message when creating a PVC with an invalid `host-storage-class` SC, we may need to validate it first. Otherwise, it will always show as bounded on the guest cluster but `not ready` on the host side.

**Review by futuretea — APPROVED** [review]
LGTM

**Comment by Vicente-Cheng** [comment]
> There is no valid message when creating a PVC with an invalid `host-storage-class` SC, we may need to validate it first. Otherwise, it will always show as bounded on the guest cluster but `not ready` on the host side.

Hi @guangbochen,

I agree with your suggestion and consideration, but the guest cluster does not have permission to check with the StorageClass of the Host Harvester Cluster. Did we need to change something config (like RBAC) then the guest cluster can have read permission to get the StorageClass list of the Host Harvester Cluster?

**Comment by guangbochen** [comment]
Adding a related `List` permission looks fine to me, thanks. @Vicente-Cheng

**Comment by Vicente-Cheng** [comment]
Hi folks, 
I update the PR and test-related works (including preparation and test cases).
Please recheck it. Thanks!

File: pkg/csi/manager.go [hunk 1]
```diff
@@ -54,6 +55,12 @@ func (m *Manager) Run(cfg *config.Config) error {
 		return err
 	}
 
+	storageClient, err := storage.NewFactoryFromConfig(restConfig)
+	if err != nil {
+		logrus.Fatalf("new storage client fail with err %v", err)
```
Comment (guangbochen) [inline pkg/csi/manager.go:60]: This might be duplicated with the returned error.

File: pkg/csi/controller_server.go [hunk 2]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
...
 }
 
+func (cs *ControllerServer) isStorageClassValid(storageClassName string) bool {
+	logrus.Infof("Prepare to check the associated StorageClass: %s", storageClassName)
+	storageClassInst, err := cs.storageClient.StorageClass().Get(storageClassName, metav1.GetOptions{})
+	if err != nil {
+		errReason := errors.ReasonForError(err)
+		switch errReason {
+		case metav1.StatusReasonForbidden:
+			logrus.Infof("No permission, skip checking. err: %+v", err)
```
Comment (guangbochen) [inline pkg/csi/controller_server.go:69]: Would prefer to use `logrus.Warnf()` here.

File: pkg/csi/controller_server.go [hunk 3]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
 	}
 }
 
+func (cs *ControllerServer) isStorageClassValid(storageClassName string) bool {
+	logrus.Infof("Prepare to check the associated StorageClass: %s", storageClassName)
```
Comment (guangbochen) [inline pkg/csi/controller_server.go:63]:
```suggestion
	logrus.Infof("Prepare to check the host StorageClass: %s", storageClassName)
```

File: pkg/csi/controller_server.go [hunk 4]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
...
+	if err != nil {
+		errReason := errors.ReasonForError(err)
+		switch errReason {
+		case metav1.StatusReasonForbidden:
+			logrus.Infof("No permission, skip checking. err: %+v", err)
+			return true
+		default:
+			logrus.Infof("Non-allowed error, fail here. err: %+v", err)
+			return false
+		}
```
Comment (guangbochen) [inline pkg/csi/controller_server.go:74]: Prefer to return `error` instead when it is not an RBAC issue.

File: pkg/csi/controller_server.go [hunk 5]
```diff
@@ -89,11 +116,30 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+	if val, exists := volumeParameters[paramHostSC]; exists {
+		targetSC = val
+	}
+	if targetSC != "" {
+		if valid := cs.isStorageClassValid(targetSC); !valid {
+			return nil, status.Error(codes.NotFound, "Associated StorageClass does not exist.")
+		}
+		pvc.Spec.StorageClassName = pointer.StringPtr(targetSC)
+		logrus.Infof("Set up the target StorageClass to : %s", *pvc.Spec.StorageClassName)
+	}
```
Comment (guangbochen) [inline pkg/csi/controller_server.go:141]: Can we move this into the `exists` block?

File: pkg/csi/controller_server.go [hunk 6]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
...
+	logrus.Infof("Prepare to check the host StorageClass: %s", storageClassName)
+	storageClassInst, err := cs.storageClient.StorageClass().Get(storageClassName, metav1.GetOptions{})
+	if err != nil {
+		errReason := errors.ReasonForError(err)
+		switch errReason {
+		case metav1.StatusReasonForbidden:
+			logrus.Warnf("No permission, skip checking. err: %+v", err)
+			return true, err
+		default:
+			logrus.Warnf("Non-allowed error, fail here. err: %+v", err)
```
Comment (futuretea) [inline pkg/csi/controller_server.go:72]: Prefer to Use logrus.Errorf

File: pkg/csi/manager.go (hunk 1 above)
Comment (Vicente-Cheng) [reply pkg/csi/manager.go]: removed, thanks!

File: pkg/csi/controller_server.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: OK, that make sense to me.

File: pkg/csi/controller_server.go [hunk 7]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
 	}
 }
 
+func (cs *ControllerServer) isStorageClassValid(storageClassName string) (bool, error) {
```
Comment (futuretea) [inline pkg/csi/controller_server.go:62]: Prefer to Use
```go
func (cs *ControllerServer) validateStorageClass(storageClassName string) error
```

File: pkg/csi/controller_server.go (hunk 3 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: updated!

File: pkg/csi/controller_server.go (hunk 4 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: done, thanks for making it more clear with errors

File: pkg/csi/controller_server.go (hunk 5 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: Done! Thanks for the tips!

File: pkg/csi/manager.go (hunk 1 above)
Comment (Vicente-Cheng) [reply pkg/csi/manager.go]: Removed!

File: pkg/csi/controller_server.go [hunk 8]
```diff
@@ -89,11 +116,31 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+	 * SC define > controller define
+	 *
+	 * Checking mechanism:
+	 * 1. If guest cluster is not be allowed to list host cluster StorageClass, just continue.
+	 * 2. If guest cluster can list host cluster StorageClass, check it.
+	 */
+	if val, exists := volumeParameters[paramHostSC]; exists {
+		//If StorageClass has `hostStorageClass` parameter, check whether it is valid or not.
+		if valid, err := cs.isStorageClassValid(targetSC); !valid {
+			return nil, status.Error(codes.Aborted, err.Error())
```
Comment (futuretea) [inline pkg/csi/controller_server.go:135]: Prefer to use codes.Internal

File: pkg/csi/controller_server.go (hunk 7 above)
Comment (futuretea) [reply pkg/csi/controller_server.go:62]: example:
```go
func (cs *ControllerServer) validateStorageClass(storageClassName string) error {
	logrus.Infof("Prepare to check the associated StorageClass: %s", storageClassName)
	sc, err := cs.storageClient.StorageClass().Get(storageClassName, metav1.GetOptions{})
	if err != nil {
		errReason := errors.ReasonForError(err)
		switch errReason {
		case metav1.StatusReasonForbidden:
			// log here
			return nil
		case metav1.StatusReasonNotFound:
			// log here
			return status.Error(codes.NotFound, " msg here")
		default:
			// log here
			return status.Error(codes.Internal, "msg here")
		}
	}
	logrus.Debugf("Get the `%s` StorageClass: %+v", storageClassName, sc)
	return nil
}
```

File: pkg/csi/controller_server.go [hunk 9]
```diff
@@ -55,8 +59,28 @@ func NewControllerServer(coreClient ctlv1.Interface, virtClient kubecli.Kubevirt
...
 }
 
+func (cs *ControllerServer) isStorageClassValid(storageClassName string) (bool, error) {
+	logrus.Infof("Prepare to check the host StorageClass: %s", storageClassName)
+	storageClassInst, err := cs.storageClient.StorageClass().Get(storageClassName, metav1.GetOptions{})
+	if err != nil {
+		errReason := errors.ReasonForError(err)
+		switch errReason {
+		case metav1.StatusReasonForbidden:
+			logrus.Warnf("No permission, skip checking. err: %+v", err)
```
Comment (bk201) [inline pkg/csi/controller_server.go:69]: It would be nice to add some comments on why it's OK to continue in this case.

Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go:73]: done

File: pkg/csi/controller_server.go (hunk 8 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: updated with the whole above suggestions, thanks!!

**Review by futuretea — APPROVED** [review]
LGTM, Thanks!

**Review by guangbochen — APPROVED** [review]
LGTM and verified that the PVC with an invalid `host-storage-class` definition will show its status and events properly.
image

**Pull request merged**